pkg/templatelookup/regular.go

```diff
@@ -239,26 +258,72 @@ func (t *TemplateLookup) getTemplate(ctx context.Context, name, desiredChannel s
 	var filteredTemplates []*v1beta2.ModuleTemplate
 	for _, template := range templateList.Items {
 		template := template
-		if template.Labels[shared.ModuleName] == name && template.Spec.Channel == desiredChannel {
+		if TemplateNameMatch(&template, name) && template.Spec.Channel == desiredChannel {
 			filteredTemplates = append(filteredTemplates, &template)
 			continue
 		}
 	}
 
+	if len(filteredTemplates) > 1 {
+		return nil, NewMoreThanOneTemplateCandidateErr(name, templateList.Items)
+	}
+
+	if len(filteredTemplates) == 0 {
+		return nil, fmt.Errorf("%w: for module %s in channel %s ",
+			ErrNoTemplatesInListResult, name, desiredChannel)
+	}
+
+	if filteredTemplates[0].Spec.Mandatory {
+		return nil, fmt.Errorf("%w: for module %s in channel %s",
+			ErrTemplateMarkedAsMandatory, name, desiredChannel)
+	}
+
+	return filteredTemplates[0], nil
+}
+
+func (t *TemplateLookup) getTemplateByVersion(ctx context.Context, name, version string) (
```

Comment:
The current implementation is fine, but we should plan a follow-up refactor to optimize the `get template` function. Instead of using `List`, we should retrieve the template directly by its `moduleTemplate` name. Since the `moduleTemplate` names follow a strict naming convention of `[module name] + [channel or version]`, it will be straightforward to implement a `Get` request for direct fetching. This change will improve performance, especially when dealing with a large number of resources, as `List` is not efficient in such scenarios.

Can you create a followup issue?

Reply:
https://github.com/kyma-project/lifecycle-manager/issues/1805